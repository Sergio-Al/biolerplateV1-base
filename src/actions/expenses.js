import { v4 as uuidv4 } from "uuid";
import db from "../firebase/firebase";
import { ref, push } from "firebase/database";

// ADD_EXPENSE
export const addExpense = (expense) => ({
  type: "ADD_EXPENSE",
  expense,
});

export const startAddExpense = (expenseData = {}) => {
  // we are returning a function that redux will exetutes
  // if is a traditional return function (not for redux) this could not work as expected
  // this is called internally by redux with dispatch as parameter (redux puts a dispatch as parameter in our return funcion)
  // this return function executes firebase operation first then dispatch to our redux store
  return (dispatch) => {
    const {
      description = "",
      note = "",
      amount = 0,
      createdAt = 0,
    } = expenseData;
    const expense = { description, note, amount, createdAt };

    return push(ref(db, "expenses"), expense).then((ref) => {
      dispatch(
        addExpense({
          id: ref.key,
          ...expense,
        })
      );
    });
  };
};

// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
  type: "REMOVE_EXPENSE",
  id,
});
// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: "EDIT_EXPENSE",
  id,
  updates,
});
