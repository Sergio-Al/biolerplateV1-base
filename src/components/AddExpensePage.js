import React from "react";

const AddExpensePage = () => <div>This is from my add expense compenent</div>;

export default AddExpensePage;
