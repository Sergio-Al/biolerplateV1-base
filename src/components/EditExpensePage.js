import React from "react";

const EditExpensePage = () => <div>This is from my Edit Page</div>;

export default EditExpensePage;
