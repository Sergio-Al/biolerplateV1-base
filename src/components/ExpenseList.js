import React from "react";
import { connect } from "react-redux";

const ExpenseList = (props) => (
  <div>
    <h1>ExpenseList</h1>
    {props.expenses.length}
    {props.filters.text}
  </div>
);

const mapStateToProps = (state) => {
  return {
    expenses: state.expenses,
    filters: state.filters,
  };
};

export default connect(mapStateToProps)(ExpenseList); // argument in connect(This of we wanna get of the store)(What we want to connect)
