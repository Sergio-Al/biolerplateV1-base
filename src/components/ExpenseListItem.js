import React from "react";
import { Link } from "react-router-dom";

const ExpenseListItem = ({ id, description, amount, createdAt }) => (
  <div>
    <h3>
      Description: {description}
      <Link to={`/edit/${id}`}> Edit</Link>
    </h3>
    <p>
      {amount} - {createdAt}
    </p>
  </div>
);

export default ExpenseListItem;
