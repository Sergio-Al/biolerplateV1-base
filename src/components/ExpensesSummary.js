import React from "react";
import { connect } from "react-redux";
import numeral from "numeral";
import expensesTotal from "../selectors/expenses-total";
import selectExpense from "../selectors/expenses";

export const ExpensesSummary = (props) => (
  <div>
    Viewing {props.expenseCount} totalling: 
    {numeral(props.expensesTotal / 100).format("$0,0.00")}
  </div>
);

const mapStateToProps = (state) => {
  return {
    expenseCount: selectExpense(state.expenses, state.filters).length,
    expensesTotal: expensesTotal(selectExpense(state.expenses, state.filters)),
  };
};

export default connect(mapStateToProps)(ExpensesSummary);
