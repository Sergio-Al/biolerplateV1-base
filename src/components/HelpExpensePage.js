import React from "react";

const HelpExpensePage = () => <div>This is from my Help Page</div>;

export default HelpExpensePage;
