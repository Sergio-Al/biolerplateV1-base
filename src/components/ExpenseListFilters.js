import React from "react";
import { connect } from "react-redux"; // to connect to the store
import { setTextFilter, sortByAmount, sortByDate } from "../actions/filters";

const ExpenseListFilters = (props) => (
  <div>
    <input
      type="text"
      value={props.filters.text}
      onChange={(e) => {
        props.dispatch(setTextFilter(e.target.value)); // we are able to change the value in the filters store
        console.log(e.target.value);
      }}
    />
    <select
      value={props.filters.sortBy}
      onChange={(e) => {
        if (e.target.value === "date") {
          props.dispatch(sortByDate());
        } else if (e.target.value === "amount") {
          props.dispatch(sortByAmount());
        }
      }}
    >
      <option value="date">Date</option>
      <option value="amount">Amount</option>
    </select>
  </div>
);

const mapStateToProps = (state) => ({
  filters: state.filters,
});
export default connect(mapStateToProps)(ExpenseListFilters); // the first argument is optional if we want to access to more store features
