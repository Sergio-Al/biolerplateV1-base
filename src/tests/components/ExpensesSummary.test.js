import React from "react";
import { shallow } from "enzyme";
import { ExpensesSummary } from "../../components/ExpensesSummary";
import expenses from "../fixtures/expenses";
import expensesTotal from "../../selectors/expenses-total";

test("should correctly render ExpensesSummary with no expenses", () => {
  const wrapper = shallow(
    <ExpensesSummary expenseCount={[].length} expensesTotal={[]} />
  );
  expect(wrapper).toMatchSnapshot();
});

test("should correctly render ExpensesSummary with 1 expense", () => {
  const wrapper = shallow(
    <ExpensesSummary expenseCount={1} expensesTotal={234} />
  );
  expect(wrapper).toMatchSnapshot();
});

test("Should correctly render ExpensesSummary with multiple expenses", () => {
  const wrapper = shallow(
    <ExpensesSummary
      expenseCount={expenses.length}
      expensesTotal={expensesTotal(expenses)}
    />
  );
  expect(wrapper).toMatchSnapshot();
});
