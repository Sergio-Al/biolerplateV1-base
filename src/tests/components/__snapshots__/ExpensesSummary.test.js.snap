// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should correctly render ExpensesSummary with multiple expenses 1`] = `
<div>
  Viewing 
  3
   
  expenses
   totalling:
  $1,141.95
</div>
`;

exports[`should correctly render ExpensesSummary with 1 expense 1`] = `
<div>
  Viewing 
  1
   
  expense
   totalling:
  $2.34
</div>
`;

exports[`should correctly render ExpensesSummary with no expenses 1`] = `
<div>
  Viewing 
  0
   
  expenses
   totalling:
  $0.00
</div>
`;
